infrastructure/metrics/src/main/java/tech/pegasys/teku/infrastructure/metrics/MetricsHistogramWithCounters.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2022 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.infrastructure.metrics;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.HdrHistogram.Histogram;
+import org.HdrHistogram.SynchronizedHistogram;
+import org.hyperledger.besu.metrics.prometheus.PrometheusMetricsSystem;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import org.hyperledger.besu.plugin.services.metrics.Counter;
+import org.hyperledger.besu.plugin.services.metrics.LabelledMetric;
+import org.hyperledger.besu.plugin.services.metrics.MetricCategory;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class MetricsHistogramWithCounters extends MetricsHistogram {
+  private final List<UInt64> sortedBoundaries;
+  private final Map<UInt64, String> boundariesToIntervalLabels;
+  private final Map<String, LabelledMetric<Counter>> intervalLabelsToMetrics;
+
+  public MetricsHistogramWithCounters(
+      final Histogram histogram,
+      final List<UInt64> sortedBoundaries,
+      final Map<UInt64, String> boundariesToIntervalLabels,
+      final Map<String, LabelledMetric<Counter>> intervalLabelsToMetrics) {
+    super(histogram);
+    this.boundariesToIntervalLabels = boundariesToIntervalLabels;
+    this.intervalLabelsToMetrics = intervalLabelsToMetrics;
+    this.sortedBoundaries = sortedBoundaries;
+  }
+
+  public static MetricsHistogramWithCounters create(
+      final MetricCategory category,
+      final MetricsSystem metricsSystem,
+      final String histogramName,
+      final String histogramHelp,
+      final String counterName,
+      final String counterHelp,
+      final int numberOfSignificantValueDigits,
+      final List<UInt64> boundaries) {
+    final List<UInt64> sortedBoundaries =
+        boundaries.stream().sorted().collect(Collectors.toUnmodifiableList());
+    final SynchronizedHistogram histogram =
+        new SynchronizedHistogram(numberOfSignificantValueDigits);
+    final Map<UInt64, String> boundariesToIntervalLabels =
+        boundariesToIntervalLabels(sortedBoundaries);
+    final Map<String, LabelledMetric<Counter>> boundariesToMetrics =
+        createCounterMetricsForBoundaries(
+            boundariesToIntervalLabels.values(), category, counterName, counterHelp, metricsSystem);
+    return createMetric(
```

Comment:
Somewhere in the setup of the metric it would be good to initialise all the labels to `0` explicitly. That way they immediately show up in the metrics output making it easy for users to see all the buckets that are available. Otherwise they'll only turn up when a value actually goes into that bucket.